src/planner/vectorfield/VectorFieldPlanner.cpp

```diff
@@ -14,229 +16,248 @@ namespace planner {
+  if (success == false)
   {
-    throw std::invalid_argument("Velocity space volume zero");
+    throw IntegrationFailedError();
   }
 
-  for (std::size_t i = 0; i < skeleton->getNumDofs(); ++i)
+  if (mEnableDofLimitCheck)
   {
-    std::stringstream ss;
-    if (skeleton->getPositionLowerLimit(i) > skeleton->getPositionUpperLimit(i))
-    {
-      ss << "Position lower limit is larger than upper limit at DOF " << i;
-      throw std::invalid_argument(ss.str());
-    }
-    if (skeleton->getVelocityLowerLimit(i) > skeleton->getVelocityUpperLimit(i))
-    {
-      ss << "Velocity lower limit is larger than upper limit at DOF " << i;
-      throw std::invalid_argument(ss.str());
-    }
+    checkDofLimits(mMetaSkeletonStateSpace, _q, _qd);
   }
+}
+
+//==============================================================================
 
-  const std::size_t numDof = _stateSpace->getDimension();
+void VectorFieldPlanner::check(const Eigen::VectorXd& _q, double _t)
+{
+  if (mTimer.getElapsedTime() > mTimelimit)
+  {
+    throw TimeLimitError();
+  }
 
-  std::vector<Knot> knots;
-  VectorFieldPlannerStatus terminationStatus
-      = VectorFieldPlannerStatus::CONTINUE;
-  std::exception_ptr terminationError;
+  // set joint values
+  mMetaSkeleton->setPositions(_q);
 
-  int cacheIndex = -1;
-  int index = 0;
-  double t = 0;
-  Eigen::VectorXd q = _stateSpace->getMetaSkeleton()->getPositions();
+  // collision checking of current joint valuesb
+  if (mEnableCollisionCheck)
+  {
+    checkCollision(mMetaSkeletonStateSpace, mConstraint);
+  }
 
-  auto startState = _stateSpace->createState();
-  _stateSpace->convertPositionsToState(q, startState);
-  Eigen::VectorXd dq(numDof);
-  assert(static_cast<std::size_t>(q.size()) == numDof);
+  Knot knot;
+  knot.mT = _t;
+  knot.mPositions = _q;
 
-  do
+  mKnots.push_back(knot);
+  mIndex += 1;
+
+  VectorFieldPlannerStatus status = mVectorField->checkPlanningStatus();
+  if (status == VectorFieldPlannerStatus::CACHE_AND_CONTINUE
+      || status == VectorFieldPlannerStatus::CACHE_AND_TERMINATE)
   {
-    // Evaluate the vector field.
-    try
-    {
-      if (!_vectorFiledCb(_stateSpace, t, dq))
-      {
-        throw VectorFieldTerminated("Failed evaluating VectorField.");
-      }
-    }
-    catch (const VectorFieldTerminated& e)
-    {
-      dtwarn << "Terminating vector field evaluation." << std::endl;
-      terminationStatus = VectorFieldPlannerStatus::TERMINATE;
-      terminationError = std::current_exception();
-      break;
-    }
+    mCacheIndex = mIndex;
+  }
 
-    if (static_cast<std::size_t>(dq.size()) != numDof)
-    {
-      throw std::length_error(
-          "Vector field returned an incorrect number of DOF velocities.");
-    }
+  if (status == VectorFieldPlannerStatus::TERMINATE
+      || status == VectorFieldPlannerStatus::CACHE_AND_TERMINATE)
+  {
+    throw VectorFieldTerminated(
+        "Planning was terminated by the StatusCallback.");
```

Comment:
Why do we treat the termination of planning (which I believe it's successfully terminated) as an exception?

Reply:
You are right. This is the way of stopping integration before it is done. Usually we set 10 sec to integrate but can reach an expected distance before that. Any other suggested way?